### DISCUSSION TITLE
Understanding LinkNeighborLoader

### DISCUSSION THREAD

**Discussion by pstrybol** [pyg-team/pytorch_geometric#7797] [Q&A] [answered]
Hi,

I am having some difficulties understanding all the arguments and return attributes of `LinkNeighborLoader` . All following questions are related to the example in this google collab: https://colab.research.google.com/drive/1r_FWLSFf9iL0OWeHeD31d_Opt031P1Nq?usp=sharing
<img width="391" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/34382674/cf11e054-5d9e-4da4-9d66-88895d48f1e4">

1. Why does the attribute `num_sampeld_nodes` equal 3 even if we specify `num_neighbors=[20, 10]` so for 2 iterations? Because the attribute `num_sampled_edges` **does** equal to 2?
1.1 does `num_neigbhors` sample 20 neigbhors for each node type and edge type?
2. When specifying `disjoint_train_ratio=0` we expect the number of edges in `edge_index` and `edge_label_index` to be the same, why is this not the case here?
3. As pointed out in discussion #7233 the attribute `node_id` is provided by the user and `n_id` is made by pyG to map the edge indices in the batch back to the original indices. But in the notebook these are the same? How can I map the batch edge indices back to the original indices? 
<img width="285" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/34382674/6c6d2169-9cf6-4d61-90a0-029126b6e5e5"> 

4. Why is has the reverse edge type less edges sampled than the normal edge type?
5. Let's say there is a second edge type e.g. 'movie', 'country_of_origin', 'country'.  If this edge type is omitted from the edge types in `RandomLinkSplit` and `LinkNeighborLoader`  is this information still used to update the embeddings of the `movie` nodes?

Thanks in advance!

**Comment by rusty1s** [comment, marked as answer]
1. There are 3 types of nodes we include when sampling for 2 hops: (1) the seed nodes, (2) the nodes in the first hop, (2) the nodes in the second hop
1.1 Yes.
2. `edge_label_index` will remove reverse edges. Might this be the case?
3. Manually setting `node_id` is no longer necessary, and is now automatically added by us via `n_id`. You can map indices back via `n_id[edge_index]`.
4. This may always happen since you will sample edges for each edge type in every hop.
5. Yes, as long as it part of the returned mini-batch, it will be used for message passing.

**Reply by pstrybol** [reply]
Thanks for your reply! I have a few follow up questions: 
1. I see. So than for the first hop we would have 20-movie nodes and 20-user nodes? Also, just to be clear with '2-hops' do you mean the second iteration of sampling first order neighbors or do you mean second order neighbors of the seed nodes. 

2. This would be strange as HeteroData specifically adds a separate reverse edge attribute. I will look into it.

Thanks

**Reply by rusty1s** [reply]
1. Yes. Regarding 2-hops, I mean the second iteration of sampling first order neighbors from sampled neighbors.
2. Keep me posted :)

**Reply by pstrybol** [reply]
Regarding 2, the `edge_label_index` also includes negative samples which are not included in the `edge_index`. I overlooked this. 

Would it be possible to define the splits yourself? Let's say I have 2 edge types and I want to train to predict 1 specific edge type. Is it as easy as just defining the `edge_label_index` and `edge_label` yourself? Also, I would not need the functionalities of `LinkNeighborLoader` so can I just use the basic `DataLoader`?

Thanks!

**Reply by rusty1s** [reply]
In general, this should be pretty straightforward. No one is forced to use `RandomLinkSplit` as it may not even fit every use-case. If you know the edges you want to predict on, use these as `edge_label_index`, and all the remaining edges as `edge_index`.

I don't see how you could use the basic `DataLoader` here though. If you are working on a single graph, you either can use no loader at all (if the whole data fits into GPU), or have to use mini-batch sampling via our neighbor loaders.

**Reply by pstrybol** [reply]
It works as expected, thanks! Maybe a little off-topic but because we use 2 different embeddings layers does is still make sense to compare distances between 2 different node types? For example, I want to see which user nodes are closest to a certain movie node in the embedding space? Can I assume that the space in which both node types are embedded are the same?

**Reply by rusty1s** [reply]
Yes, if you train the embeddings via inner product, this is not a problem.

**Comment by songsong0425** [comment]
Hi, sorry for updating the completed question. Does the numbers in `num_neighbors=[n_1, ..., n_k]` mean that sample the k-hop neighbors from seed nodes? Until now, I thought that n_k meant the number of neighbors per source and target node. Then, if I run the 3-layered GCN, should I set the `num_neighbors=[n_1, n_2, n_3]?`

**Reply by rusty1s** [reply]
`num_neighbors=[n_1, n_2, n_3]` means that you sample `n_1` neighbors for every seed node in the first hop, you sample `n_2` neighbors in the second hop for each neighbor sampled in the first hop, and you sample `n_3` neighbors in the third hop for each neighbor sampled in the second hop.

**Reply by songsong0425** [reply]
I understand. Thank you for your kind reply!
